trieste/acquisition/function.py

```diff
@@ -509,6 +512,155 @@ def prepare_acquisition_function(
+        )
+
+        def Psi(a, b, mean, std) -> TensorType:
+            """
+            Generic Expected Improvement defined at Eq. 19 in :cite:`yang2019efficient`,
+            calculate the expected improvement on reference a, integrating from b
+            :param a: shape [num_cells, out_dim] expected improvement reference point
+            :param b: shape [num_cells, out_dim] integral upper bounds
+            :param mean: shape [..., out_dim]
+            :param std: shape [..., out_dim]
+            """
+            return std * normal.prob((b - mean) / std) + (mean - a) * (
+                1 - normal.cdf((b - mean) / std)
+            )
+
+        def nu(lb, ub, mean, std) -> TensorType:
+            """
+            A special expected improvement given fixed improvement,
+            defined at Eq. 25 in :cite:`yang2019efficient`
+            :param lb: shape [..., num_cells, out_dim] slice lower bound
+            :param ub: shape [..., num_cells, out_dim] slice upper bound
+            :param mean: shape [..., out_dim]
+            :param std: shape [..., out_dim]
+            """
+            return (ub - lb) * (1 - normal.cdf((ub - mean) / std))
+
+        candidate_mean, candidate_var = model.predict(tf.squeeze(x, -2))
+        candidate_std = tf.sqrt(candidate_var)
+
+        # calc ehvi assuming maximization
+        neg_candidate_mean = -tf.expand_dims(candidate_mean, 1)  # [..., 1, out_dim]
+        candidate_std = tf.expand_dims(candidate_std, 1)  # [..., 1, out_dim]
+
+        lb_points, ub_points = pareto.hypercell_bounds(
+            tf.constant([-inf] * candidate_mean.shape[-1], dtype=x.dtype), reference_point
+        )
+
+        neg_lb_points, neg_ub_points = -ub_points, -lb_points
+
+        neg_ub_points = tf.minimum(
+            neg_ub_points, 1e10
+        )  # this maximum: 1e10 is heuristically chosen
+
+        psi_ub = Psi(
+            neg_lb_points, neg_ub_points, neg_candidate_mean, candidate_std
+        )  # [..., num_cells, out_dim]
+        psi_lb = Psi(
+            neg_lb_points, neg_lb_points, neg_candidate_mean, candidate_std
+        )  # [..., num_cells, out_dim]
+
+        psi_lb2ub = tf.maximum(psi_lb - psi_ub, 0.0)  # [..., num_cells, out_dim]
+        nu_contrib = nu(neg_lb_points, neg_ub_points, neg_candidate_mean, candidate_std)
+
+        # get stacked factors of Eq. 45
+        # [2^m, indices_at_dim]
+        cross_index = tf.constant(list(product(*[[0, 1]] * reference_point.shape[-1])))
+
+        # Take the cross product of psi_diff and nu across all outcomes
+        # [..., num_cells, 2(operation_num, refer Eq. 45), num_obj]
+        stacked_factors = tf.concat(
+            [tf.expand_dims(psi_lb2ub, -2), tf.expand_dims(nu_contrib, -2)], axis=-2
+        )
+
+        # [..., num_cells, 2^m, 2(operation_num), num_obj]
+        factor_combinations = tf.linalg.diag_part(tf.gather(stacked_factors, cross_index, axis=-2))
+
```

Comment:
Could lines 632-643 (and the long operation in the return statement of 647) be made into a function? 

You could then define this function with an inline comment  describing what it is does (i.e eq. 45 ...)

See line 210 for MinValueEntropySearch, where we keep the code interpretable by defining functions.

Reply:
Rearranged into a new function (start at line 604).